arch/x86_64: Fix TLS alignment issue

cffs opened the pull request:
According to the x86_64 ELF ABI, the thread pointer in FS register should be aligned according to the alignment of the TLS program header in the ELF file. As this pointer should be just after the TLS area, the TLS area itself has to be properly aligned and padded.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://docs.unikraft.org/contribute.html) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): x86_64
 - Platform(s): N/A
 - Application(s): N/A

### Description of changes

Change the TLS area setup so that the thread pointer always has sufficient (i.e. 32-byte) alignment. This amounts to increase alignment and padding of the TLS area. Additionally, we ensure there is always at least one variable in `.tbss` section, to work around a ld limitation (ld ignores the `.tbss` section if it contains only padding).

mogasergiu commented on arch/x86/x86_64/include/uk/asm/tls.h:
```diff
@@ -67,5 +81,8 @@ static inline void ukarch_tls_area_copy(void *tls_area)
+	__sz tls_len = _tls_end - _tls_start;
+	__sz padding = tls_len % ukarch_tls_area_align();
+
+	return tls_area + padding + tls_len;
```
I believe it should have been `return tls_area + ukarch_tls_area_size();`. Otherwise `fs:offset` seems to pick wrong values.

cffs replied:
`ukarch_tls_area_size()` returns the full TLS area size, including **8 bytes for the TLS pointer itself** appended at the end of it. It does seem to me the TLS pointer address should be `tls_area + ukarch_tls_area_size() - 8`, which is the same as `tls_area + padding + tls_len`. It was already the case before my changes, which amounted to add the padding both to `ukarch_tls_area_size()` and to `ukarch_tls_pointer()`.
Can you provide a [MWE](https://www.wikiwand.com/en/Minimal_working_example) that triggers wrong values for `fs::offset`, so I can have a look? It seems to work on my setup.
That being said, I think returning `tls_area + ukarch_tls_area_size() - 8` would be clearer and less redundant than what is currently present. I'll update the PR with that version.

cffs replied:
I found some issue on my own, I will investigate and come back to you.

mogasergiu replied:
I have provided [here](https://github.com/cs-pub-ro/app-helloworld/tree/TLS-alignment) the example which fails on my end. With this commit, the output is `0x0.0x0.0x0`, instead of `0x22.0x4.0x4`.

cffs replied:
There is some problem unrelated to this patch, that also occurs with current staging. I created a new issue: #290. Let's revisit the alignment issue once #290 is fixed.

cffs committed:
arch/x86_64: increase TLS alignment

Some CPU instructions require the operand to be aligned on a 16-byte
or 32-byte boundary. Such operations may involved thread-local
variables. When it is the case, the compiler (and linker) will increase
the alignment requirements for the `.tdata` and `.tbss` sections in the
ELF objects, and in TLS program header.

As Unikraft `uksched` TLS handling currently ignores alignment
requirements set in the ELF image, this commit increases the constant
alignment from 8 to 32, to allow for all use cases.

Signed-off-by: Cyril Soldani <cyril.soldani@uliege.be>

cffs commented:
I found a simpler way to fix the alignment issue than in the original commit, by asking for more alignment in the linker script rather than doing manual padding in the code. However, there is still issue #290 when the `.tbss` section should be missing, but is not.

mogasergiu reviewed (approved):
Everything looks fine now.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

cffs committed:
arch/x86_64: add dummy thread-local variable

This ensures there will be at least one variable in the `.tbss` section.
If that section is empty and only contains padding for alignment
purpose, then TLS offsets may be wrong.

Closes #290.

Note that is would not be necessary if there was no padding added to the
`.tbss` section by the linker script, but x86_64 TLS code would then
have to be updated to read the actual required alignment from the ELF
file. Using a fixed alignment is much simpler to implement, and the
introduction of this thread-local variable should only cost 4 additional
bytes per thread.

Signed-off-by: Cyril Soldani <cyril.soldani@uliege.be>

cffs commented:
Rather than adding a thread-local variable in `uksched` (which calls TLS initialisation functions), I added it in the x86_64 architecture code, since it should only be necessary on that architecture.

vladandrew commented:
Approved-by: Vlad-Andrei Badoiu <vlad_andrei.badoiu@upb.ro>
